src/app/api/game-sessions/route.ts

```diff
@@ -20,21 +91,280 @@ export async function GET() {
 
-    if (new Date(body.date) < new Date()) {
+    if (new Date(gameSessionToInsert.date) < new Date()) {
       return new Response("Date cannot be in the past", { status: 400 });
     }
 
-    await db.insert(gameSessions).values(body);
+    // Check that a record doesn't already exist for this date
+    const gameSession = await db.query.gameSessions.findFirst({
+      where: eq(gameSessions.date, gameSessionToInsert.date),
+    });
+
+    if (gameSession) {
+      return new Response("A game session already exists for this date", {
+        status: 400,
+      });
+    }
+
+    // Check for a schedule on this weekday
+    const semester = await db.query.semesters.findFirst({
+      with: {
+        gameSessionSchedules: {
+          where: eq(
+            gameSessionSchedules.weekday,
+            getWeekday(gameSessionToInsert.date),
+          ),
+        },
+      },
+      where: or(
+        and(
+          lte(semesters.startDate, gameSessionToInsert.date),
+          gt(semesters.breakStart, gameSessionToInsert.date),
+        ),
+        and(
+          lt(semesters.breakEnd, gameSessionToInsert.date),
+          gte(semesters.endDate, gameSessionToInsert.date),
+        ),
+      ),
+    });
+
+    const gameSessionSchedule = semester?.gameSessionSchedules[0];
+
+    if (gameSessionSchedule) {
+      // If a schedule exists, only create record if an exception also exists, otherwise it will be generated
+      const gameSessionException =
+        await db.query.gameSessionExceptions.findFirst({
+          where: eq(
+            gameSessionExceptions.gameSessionDate,
+            gameSessionToInsert.date,
+          ),
+        });
 
-    return NextResponse.json(body, { status: 201 });
+      if (!gameSessionException) {
+        return new Response("No gameSessionException found for this date", {
+          status: 400,
+        });
+      }
+    }
+
+    // Check that the gameSession is valid
+    if (gameSessionToInsert.bookingOpen > new Date(gameSessionToInsert.date)) {
+      return new Response("Booking open must be before the game session date", {
+        status: 400,
+      });
+    }
+
+    if (gameSessionToInsert.bookingClose > new Date(gameSessionToInsert.date)) {
+      return new Response(
+        "Booking close must be before the game session date",
+        {
+          status: 400,
+        },
+      );
+    }
+
+    if (gameSessionToInsert.bookingOpen > gameSessionToInsert.bookingClose) {
+      return new Response("Booking open must be before booking close", {
+        status: 400,
+      });
+    }
+
+    if (gameSessionToInsert.startTime > gameSessionToInsert.endTime) {
+      return new Response("Start time must be before end time", {
+        status: 400,
+      });
+    }
+
+    if (gameSessionToInsert.casualCapacity > gameSessionToInsert.capacity) {
+      return new Response("Casual capacity must be less than capacity", {
+        status: 400,
+      });
+    }
```

Comment:
These checks should be put towards the start of the function so we don't make any unnecessary db calls